sympy/integrals/risch.py

```diff
@@ -218,14 +218,20 @@ def __init__(self, f=None, x=None, handle_first='log', dummy=False, extension=No
         self.dummy = dummy
         self.reset()
         exp_new_extension, log_new_extension = True, True
+
+        # case of 'automatic' choosing
+        if rewrite_complex is None:
+            rewrite_complex = I in self.f.atoms()
+
         if rewrite_complex:
             rewritables = {
                 (sin, cos, cot, tan, sinh, cosh, coth, tanh): exp,
                 (asin, acos, acot, atan): log,
             }
-        # rewrite the trigonometric components
+            # rewrite the trigonometric components
             for candidates, rule in rewritables.items():
                 self.newf = self.newf.rewrite(candidates, rule)
+            self.newf = cancel(self.newf)
```

Comment:
I don't think we will be needing two separate (one each after `exp` and `log` rewrite) or just one after both the rewrites have been done.

Reply:
Currently it is failing a test https://travis-ci.org/sympy/sympy/jobs/270698759 .https://github.com/sympy/sympy/blob/96074807d8d003dbd83198d82ee0d5b16390709d/sympy/integrals/tests/test_integrals.py#L897 . Though risch algorithm is not able (non-elementary) integrate it but it modifies the expression for `risch=False` as well, and that is a problem for it.

Reply:
Can that related test be changed to `risch=False`?

Reply:
How does it fail the test if risch cannot integrate it?

Reply:
The risch modifies the expression to separate it (into `exp(-x**2)*exp(I*c*x)`), that is being done because of `cancel` function.

This is similar to passing `integrate(exp(-x**2)*exp(I*c*x), x)` into current function.

Reply:
Does the risch rewriting affect the other integrate routines? 

Anyway, how is the test failing? If it's still the right answer in a different form that's fine.

Reply:
It is returned as it is:
```
In [3]: integrate(exp(-x**2 + I*c*x), x)
Out[3]: 
⌠               
⎮    2          
⎮  -x   ⅈ⋅c⋅x   
⎮ ℯ   ⋅ℯ      dx
⌡               
```

Reply:
> Does the risch rewriting affect the other integrate routines?

Yes, I think it does affect.

Reply:
So let's make it so that if the entire expression is nonelementary that it reuses the original expression, not the rewritten one.